alibaba/DeepRec
# [Embedding] Add experimental libpmem based PMEM allocator

## Architecture

Please ref to [Doc](https://deeprec.yuque.com/deeprec/wvzhaq/vom2mu)

## How to test

- install dependencies: scripts/install_pmem_dependencies.sh
- mount PMEM with FSDAX mode
- Modify "alloc_" at line 173 of tensorflow/core/framework/embedding/embedding_var.h with cpu_allocator(). (this is a known issue of master)
- In ./configure, choose y for build TensorFlow with PMEM support
- Build
- run launch.sh with  "--ev_storage=pmem_libpmem --ev_storage_path=<path to mounted pmem (default: /mnt/pmem0/allocator)> --ev_storage_size_gb=<size in gb (default: 512)>"
- to get best performance, bind the same socket for cpu and pmem (e.g. numactl)

Notice: the internal configuration of the allocator is hardcoded at the top of experimental_pmem_allocator.h, you can test with the default configuration, or modify them ref to [Doc](https://deeprec.yuque.com/deeprec/wvzhaq/vom2mu)
